### DISCUSSION TITLE
tutorials/inputs/inputevent

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#19] [User-contributed notes]
# tutorials/inputs/inputevent

What is it?: Managing input is usually complex, no matter the OS or platform. To ease this a little, a special built-in type is provided, InputEvent. This datatype can be configured to contain seve...

https://docs.godotengine.org/en/latest/tutorials/inputs/inputevent.html

**Comment by enetheru** [comment]
Under the heading 'How does it work?' points 6 and 7 are mostly duplicates.

**Reply by GeminiSquishGames** [reply]
5 is "Mostly" the same too, but each of the 3 describe a different method from each other. "Mostly" sure, but each different enough to explain each method with caveats from the rest. Explicitly declare your documentation and your scripts, yo. Best practice.

**Reply by jose24342801-ops** [reply]
67 so funny

**Comment by DaltonSW** [comment]
Is there a reason that the recommendation is to check `IsActionPressed` in `_Process` instead of `_UnhandledInput`? The docs are worded like you must check `@event` for info. Seems like it could be more efficient to do your various action checks in the input-triggered function so they're not running every frame

**Reply by GeminiSquishGames** [reply]
It says, "However, it is cleaner and more flexible to use the provided InputMap feature..." and explains how to do do it arbitrarily with the Input class. The check is pretty negligible and IIRC event is populated every frame anyway by the DisplayServer class, so no savings happen under the hood, so the check is the same regardless if you do it in the game loop or one of the input process events, it's just checking the already populated event either way each frame. Also, Input.event has the methods ...JustPressed, ...JustReleased, and ...RawStrength for actions, which again IIRC don't work on the @event object inside of input process events.

**Reply by XandaPanda42** [reply]
I remember getting confused by that, thinking that meant that I should prefer to use the Input singleton *instead* of using `_unhandled_input()`.

If the difference is negligible, maybe we should opt for consistency instead? Having the second example be:

```gdscript
func _unhandled_input(event):
    if event.is_action_pressed("ui_cancel"):
        get_tree().quit()
```

partly so that the example showcases the Input action class doing the same thing as the original (exiting the game when you press escape), but secondly so it uses the InputEvent class, instead of the Input singleton. 

I think having the example be in the _process function just muddies things, and doesn't use the InputEvent class which is what this tutorial is about.

Unless that part is saying that it's better to use the Input class in _process instead of using InputEvents at all, in which case that could probably be made clearer.

**Comment by jgedri** [comment]
If this is a tutorial, it spends little time on explaining how to do things.

**Reply by Calinou** [reply]
Please open issues on https://github.com/godotengine/godot-docs to request documentation improvements, and be specific about what needs to be changed.

**Reply by HalfDeafDev** [reply]
I don't believe we should consider this a tutorial, this is documentation on using InputEvent.

Tutorials take you from point A to point B, normally through hand holding, that keeps the messy details away.

Documentation gives brief examples without context and exposes you to messy details.

**Comment by zmra** [comment]
I found this video to be helpful in untangling inputevents and input handling in general:  https://www.youtube.com/watch?v=AfX00eQIRCU

**Comment by hungptsg** [comment]
The docs should mentioned that for a Control, `_gui_input` will only report key events if control is focused.

Also for `_unhandled_input` to report mouse, parent and control `mouse_filter` must be `pass`.

**Comment by ITamas78** [comment]
Hello!
How can I detect two touch events simultaneously on the screen of an Android device? My code works if event position smaller than position of the tank node, but I can't shoot when I move right with the tank.

```gdscript
func _input(event):
    if event is InputEventScreenTouch:
        if event.pressed:
            if fingeroncrosshair(event.position.x,event.position.y)==true:
                canfire=true #Shoot!
            
            if event.position.x<position.x:
                tankmovedirection=1 #Go left
            else:
                tankmovedirection=2 #Go right
        else:
            tankmovedirection=0 #Stop!
            canfire=false #Don't shoot
    
    if event is InputEventScreenDrag:
        if fingeroncrosshair(event.position.x,event.position.y)==true:
            if event.position.x<position.x: #Crosshair follows player's finger
                get_node("../crosshair").position=event.position
```

**Comment by XandaPanda42** [comment]
When we use an InputEvent in _unhandled_input, if we use that input for movement or control or whatever, should we be calling Viewport.set_input_as_handled() to stop every other node from receiving that event and doing every check in their _unhandled_input as well? If we don't want it to be used for something else as well obviously.

And how does that work? It doesn't let us pass the event as an argument or anything, so does it just set the last registered event as "handled"? Does it just stop every other node after it from running the unhandled_input function and skip to the next part of the code? I'm a little unclear sorry.
